src/lib/tuning/tuner.cpp

```diff
@@ -0,0 +1,197 @@
+}
+
+float Tuner::time_budget() const { return _time_budget.count(); }
+
+float Tuner::evaluate_time_budget() const { return _evaluate_time_budget.count(); }
+
+float Tuner::select_time_budget() const { return _select_time_budget.count(); }
+
+float Tuner::execute_time_budget() const { return _execute_time_budget.count(); }
+
+void Tuner::set_cost_budget(float budget) { _cost_budget = budget; }
+
+float Tuner::cost_budget() const { return _cost_budget; }
+
+void Tuner::schedule_tuning_process() {
+  Assert(_evaluators.size() > 0, "Can not run Tuner without at least one AbstractEvaluator");
+  Assert(_selector, "Can not run Tuner without an AbstractSelector");
+  Assert(!is_running(), "Can not schedule another tuning process while the previous process is still running");
+
+  _evaluate_task = std::make_shared<JobTask>([this]() { this->_evaluate(); });
+  _select_task = std::make_shared<JobTask>([this]() { this->_select(); });
+  _execute_task = std::make_shared<JobTask>([this]() { this->_execute(); });
+
+  _evaluate_task->set_as_predecessor_of(_select_task);
+  _select_task->set_as_predecessor_of(_execute_task);
+
+  _status = Status::Running;
+  _remaining_time_budget = _time_budget;
+  _time_budget_exceeded = false;
+
+  _evaluate_task->schedule();
+  _select_task->schedule();
+  _execute_task->schedule();
+}
+
+bool Tuner::is_running() {
+  return (_evaluate_task && !_evaluate_task->is_done()) || (_select_task && !_select_task->is_done()) ||
+         (_execute_task && !_execute_task->is_done());
+}
+
+Tuner::Status Tuner::status() { return _status; }
+
+void Tuner::wait_for_completion() {
+  if (_evaluate_task) {
+    _evaluate_task->join();
+  }
+  if (_select_task) {
+    _select_task->join();
+  }
+  if (_execute_task) {
+    _execute_task->join();
+  }
+}
+
+void Tuner::_evaluate() {
+  LOG_INFO("Begin tuning evaluation phase...");
+
+  auto begin = RuntimeClock::now();
+
+  _choices.clear();
+  for (const auto& evaluator : _evaluators) {
+    evaluator->evaluate(_choices);
+
+    auto runtime = RuntimeClock::now() - begin;
+    if (runtime > _remaining_time_budget) {
+      LOG_INFO("Interrupt evaluation phase: Time budget exceeded.");
+      _status = Status::Timeout;
+      return;
+    }
+    if (runtime > _evaluate_time_budget) {
+      LOG_INFO("Interrupt evaluation phase: Evaluate time budget exceeded.");
+      _status = Status::EvaluationTimeout;
+      return;
+    }
+  }
+  _remaining_time_budget -= RuntimeClock::now() - begin;
+
+  _log_choices();
+  LOG_INFO("Tuning evaluation phase completed.");
+}
+
+void Tuner::_select() {
+  if (_status != Status::Running) {
+    LOG_INFO("Skip tuning selection phase because previous step exceeded the time budget.");
+    return;
+  }
+  LOG_INFO("Begin tuning selection phase...");
+
+  auto begin = RuntimeClock::now();
+
+  _operations = _selector->select(_choices, _cost_budget);
+
+  auto runtime = RuntimeClock::now() - begin;
+  if (runtime > _remaining_time_budget) {
```

Comment:
Does this mean that we break here even though the selector has finished? We break because he has used more time than planned?

Reply:
Exactly.